rust/xaynet-core/src/mask/masking.rs

```diff
@@ -585,6 +587,155 @@ mod tests {
     test_masking!(pow_i64_b6, Power2, i64, 1_000_000, 10);
     test_masking!(pow_i64_bmax, Power2, i64, 10);
 
+    /// Generate tests for masking and unmasking of a single model:
+    /// - generate random scalar from a uniform distribution with a seeded PRNG
+    /// - scale a model of unit weights and mask it
+    /// - check that all masked weights belong to the chosen finite group
+    /// - unmask the masked model
+    /// - check that all unmasked weights are equal to the original weights (up to a tolerance
+    ///   determined by the masking configuration)
+    ///
+    /// The arguments to the macro are:
+    /// - a suffix for the test name
+    /// - the group type of the model (variants of `GroupType`)
+    /// - the data type of the model (either primitives or variants of `DataType`)
+    /// - an absolute bound for the weights (optional, choices: 1, 100, 10_000, 1_000_000)
+    /// - the number of weights
+    macro_rules! test_masking_scalar {
+        ($suffix:ident, $group:ty, $data:ty, $bound:expr, $len:expr $(,)?) => {
+            paste::item! {
+                #[test]
+                fn [<test_masking_scalar_ $suffix>]() {
+                    // Step 1: Build the masking config
+                    let config = MaskConfig {
+                        group_type: $group,
+                        data_type: paste::expr! { [<$data:upper>] },
+                        bound_type: match $bound {
+                            1 => B0,
+                            100 => B2,
+                            10_000 => B4,
+                            1_000_000 => B6,
+                            _ => Bmax,
+                        },
+                        model_type: M3,
+                    };
+                    let vect_len = $len as usize;
+
+                    // Step 2: Generate a random scalar from (0, bound]
+                    // take vector [1, ..., 1] as the model to scale
+                    let bound = if $bound == 0 {
+                        paste::expr! { [<$data:lower>]::MAX / (2 as [<$data:lower>]) }
+                    } else {
+                        paste::expr! { $bound as [<$data:lower>] }
+                    };
+                    let zero = 0 as [<$data:lower>];
+                    let mut prng = ChaCha20Rng::from_seed(MaskSeed::generate().as_array());
+                    let rand_neg = Uniform::new(-bound, zero).sample(&mut prng);
+                    let scalar = rand_neg.abs() as f64;
```

Comment:
to avoid the double negation of the scalar and to still sample it from the correct interval, this could also be written as
```suggestion
                    let eps = [<$data:lower>]::EPSILON;
                    let mut prng = ChaCha20Rng::from_seed(MaskSeed::generate().as_array());
                    let scalar = Uniform::new_inclusive(eps, bound).sample(&mut prng) as f64;
```

Reply:
nice suggestion :+1: